pkg/repro/repro.go

```diff
@@ -168,46 +243,212 @@ func (ctx *context) reproExtractProg(entries []*prog.LogEntry) (*Result, error)
+			for _, prog := range c {
+				progs = append(progs, prog)
+			}
+		}
+		progs = append(progs, chunk1...)
+		progs = append(progs, chunk2...)
+		for _, c := range guilty2 {
+			for _, prog := range c {
+				progs = append(progs, prog)
+			}
+		}
+		return progs
+	}
+
+	logGuilty := func(guilty [][]*prog.LogEntry) string {
+		log := "["
+		for i, chunk := range guilty {
+			log += fmt.Sprintf("<%d>", len(chunk))
+			if i != len(guilty)-1 {
+				log += ", "
+			}
+		}
+		log += "]"
+		return log
+	}
+
+	// Bisect the programs to find the ones that cause the crash.
+	guilty := [][]*prog.LogEntry{entries}
+again:
+	Logf(3, "reproducing crash '%v': bisect: guilty chunks: %v", ctx.crashDesc, logGuilty(guilty))
+	for i, chunk := range guilty {
+		if len(chunk) == 1 {
+			continue
+		}
+
+		guilty1 := guilty[:i]
+		guilty2 := guilty[i+1:]
+		Logf(3, "reproducing crash '%v': bisect: guilty chunks split: %v, <%v>, %v", ctx.crashDesc, logGuilty(guilty1), len(chunk), logGuilty(guilty2))
+
+		chunk1 := chunk[0 : len(chunk)/2]
+		chunk2 := chunk[len(chunk)/2 : len(chunk)]
+		Logf(3, "reproducing crash '%v': bisect: chunk split: <%v> => <%v>, <%v>", ctx.crashDesc, len(chunk), len(chunk1), len(chunk2))
+
+		Logf(3, "reproducing crash '%v': bisect: triggering crash without chunk #1", ctx.crashDesc)
+		progs := compose(guilty1, guilty2, []*prog.LogEntry{}, chunk2)
+		crashed, err := ctx.testProgs(progs, duration(len(progs)), opts)
+		if err != nil {
+			return nil, err
+		}
+
+		if crashed {
+			guilty = [][]*prog.LogEntry{}
+			guilty = append(guilty, guilty1...)
+			guilty = append(guilty, chunk2)
+			guilty = append(guilty, guilty2...)
+			Logf(3, "reproducing crash '%v': bisect: crashed, chunk #1 evicted", ctx.crashDesc)
+			goto again
+		}
+
+		Logf(3, "reproducing crash '%v': bisect: triggering crash without chunk #2", ctx.crashDesc)
+		progs = compose(guilty1, guilty2, chunk1, []*prog.LogEntry{})
+		crashed, err = ctx.testProgs(progs, duration(len(progs)), opts)
+		if err != nil {
+			return nil, err
+		}
+
+		if crashed {
+			guilty = [][]*prog.LogEntry{}
+			guilty = append(guilty, guilty1...)
+			guilty = append(guilty, chunk1)
+			guilty = append(guilty, guilty2...)
+			Logf(3, "reproducing crash '%v': bisect: crashed, chunk #2 evicted", ctx.crashDesc)
+			goto again
+		}
+
+		guilty = [][]*prog.LogEntry{}
+		guilty = append(guilty, guilty1...)
+		guilty = append(guilty, chunk1)
+		guilty = append(guilty, chunk2)
+		guilty = append(guilty, guilty2...)
+
+		Logf(3, "reproducing crash '%v': bisect: not crashed, both chunks required", ctx.crashDesc)
+
+		goto again
+	}
+
+	// Concatenate all programs into one.
+	entries = []*prog.LogEntry{}
+	var prog prog.Prog
+	for _, chunk := range guilty {
+		if len(chunk) != 1 {
+			return nil, fmt.Errorf("bad bisect result: %v", guilty)
+		}
+		entries = append(entries, chunk[0])
+		prog.Calls = append(prog.Calls, chunk[0].P.Calls...)
+	}
+
+	// TODO: Minimize each program before concatenation.
```

Comment:
Try removing calls from each program separately, but execute them together to test whether they still cause a crash.